decouple QML from legacy skin system + add preferences button

Be-ing opened the pull request:
Somehow this is hanging in the ControllerManager destructor. It seems the `requestShutdown` signal does not trigger `slotShutdown` so `m_pThread->wait()` blocks indefinitely.

Be-ing committed:
add QmlApplication class activated by --qml CLI option

uklotzde commented on src/qmlapplication.h:
```diff
@@ -0,0 +1,38 @@
+
+  private:
+    std::shared_ptr<CoreServices> m_pCoreServices;
+    const CmdlineArgs& m_cmdlineArgs;
```
We should not store a plain reference to CmdlineArgs as a member and instead obtain it with CmdlineArgs::Instance(). This is already wrong in CoreServices, I could submit a PR to fix that.

Be-ing replied:
CmdlineArgs was unused copypasta from #3448 anyway. Removed and force pushed.

uklotzde commented on src/qmlapplication.h:
```diff
@@ -0,0 +1,38 @@
+    QmlApplication(
+            QApplication* app,
+            std::shared_ptr<CoreServices> pCoreServices,
+            const CmdlineArgs& args);
```
CmdlineArgs is a singleton and does not need to be passed around -> remove

Be-ing commented:
Any ideas why ~ControllerManager is hanging?

uklotzde commented:
> Any ideas why ~ControllerManager is hanging?

I didn't have time to test it yet. Will do later.

Be-ing committed:
remove QmlSkin class

Be-ing committed:
move src/skin/qml to src/qml

Be-ing committed:
move mixxx::skin::qml namespace to mixxx::qml

Be-ing committed:
move src/qmlapplication.{cpp,h} to src/qml

Be-ing committed:
MixxxApplication: remove Qt::AA_DontCreateNativeWidgetSiblings

That was required for hacking QML into the legacy skin system
but it created a regression with the cursor getting stuck on the
splitter dragging handle:
https://bugs.launchpad.net/mixxx/+bug/1943860

Be-ing committed:
move res/skins/QMLDemo to res/qml

Be-ing commented:
Effects including EQs are not working with QML. I tested merging #2618 but that does not fix it. There must be some implicit dependency of the effects system on the setup of the legacy GUI.

Be-ing commented:
Well this is really ugly. Initialization of the effects system depends on the existence of a DlgPreferences instance.

Be-ing commented:
Even uglier: the hang in ~ControllerManager goes away when a DlgPreferences instance is created.

uklotzde commented on src/qml/qmlapplication.h:
```diff
@@ -0,0 +1,33 @@
+    QmlApplication(
+            QApplication* app,
+            std::shared_ptr<CoreServices> pCoreServices);
+    ~QmlApplication() = default;
```
`override`

Be-ing replied:
default override? Would it be better to simply remove this line?

uklotzde replied:
No, `override` should be used for all destructors of derived classes as a convention. C++ is so error prone and requires a lot of informal boilerplate.

uklotzde replied:
If omitted the base class is not checked for a virtual destructor. QObject has one, but nevertheless follow the convention. If you inherit than add either a `virtual` or an `override` destructor.

Be-ing replied:
Fixed and squashed into the first commit.

uklotzde commented on src/qml/qmlapplication.h:
```diff
@@ -0,0 +1,33 @@
+  private:
+    std::shared_ptr<CoreServices> m_pCoreServices;
+
+    QString m_mainQmlFilePath;
```
`m_rootQmlFilePath`

Be-ing replied:
The QML file is named `main.qml` so I think this variable name is appropriate.

uklotzde commented on src/qml/qmlapplication.h:
```diff
@@ -0,0 +1,33 @@
+    QString m_mainQmlFilePath;
+
+    std::unique_ptr<QQmlApplicationEngine> m_pQmlAppEngine;
+    QFileSystemWatcher m_qmlFileWatcher;
```
Shall we drop the redudndant "qml/""Qml" term from all members? It seems to be redundant. This class should not take more responsibilities so it might to be safe.

Be-ing replied:
Yes, good idea.

Be-ing replied:
Should remove the `Qml` prefix from the class names within the `mixxx::qml` namespace?

uklotzde replied:
We could decide about the naming later. Let's leave it as is until we are sure.

uklotzde reviewed (commented):
Dropping the "skin" prefix from namespaces and file paths is a good idea. We have *skins* and we have *qml*, mutually exclusive. This avoids to distinguish between *legacy skins* and *QML skins*.

Be-ing commented:
More ugliness: I cannot include include preferences/dialog/dlgpreferences.h in src/qml/qmlapplication.h because compiling main.cpp fails with:
```
In file included from ../src/qml/qmlapplication.h:4,
                 from ../src/main.cpp:14:
../src/preferences/dialog/dlgpreferences.h:13:10: fatal error: preferences/dialog/ui_dlgpreferencesdlg.h: No such file or directory
   13 | #include "preferences/dialog/ui_dlgpreferencesdlg.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

I tried adding

```
add_executable(mixxx WIN32 src/main.cpp src/qml/qmlapplication.cpp)
set_target_properties(mixxx PROPERTIES AUTOUIC ON AUTOMOC ON)
```
to CMakeLists.txt and removing src/qml/qmlapplication.cpp from the sources list for the mixxx-lib target. Unfortunately the build still fails. CMake is generating ui_dlgpreferencesdlg.h but it is located in build/mixxx-lib_autogen/include/preferences/dialog/ui_dlgpreferencesdlg.h which is not in the list of include paths for the mixxx executable target. Any suggestions how to resolve this?

Be-ing commented:
I made a QML button to show the preferences but I am puzzled how closing the preferences window triggers the QmlApplication destructor. I set a breakpoint in ~DlgPreferences which shows that DlgPreferences is properly getting deleted by QmlApplication... but why is QmlApplication getting destroyed in the first place when DlgPreferences is closed?? It does not matter whether DlgPreferences is closed by the Okay, Apply, or Cancel button or the window manager close button. I commented out adding all the pages in the DlgPreferences constructor but I still reproduced the issue. I also tried creating a QQuickView instead of a QQmlApplicationEngine but the issue still occurred.

```
0  DlgPreferences::~DlgPreferences() (this=0x9c1d060, __in_chrg=<optimized out>) at ../src/preferences/dialog/dlgpreferences.cpp:297
#1  0x000000000059394a in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release() (this=0x9c1d050)
    at /usr/include/c++/11/bits/shared_ptr_base.h:168
#2  std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release() (this=0x9c1d050) at /usr/include/c++/11/bits/shared_ptr_base.h:161
#3  std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count() (this=0x7fffffffdc48, __in_chrg=<optimized out>)
    at /usr/include/c++/11/bits/shared_ptr_base.h:705
#4  std::__shared_ptr<DlgPreferences, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr() (this=0x7fffffffdc40, __in_chrg=<optimized out>)
    at /usr/include/c++/11/bits/shared_ptr_base.h:1154
#5  std::shared_ptr<DlgPreferences>::~shared_ptr() (this=0x7fffffffdc40, __in_chrg=<optimized out>) at /usr/include/c++/11/bits/shared_ptr.h:122
#6  mixxx::qml::QmlApplication::~QmlApplication() (this=0x7fffffffdc00, __in_chrg=<optimized out>) at ../src/qml/qmlapplication.h:19
#7  0x00000000005740c9 in (anonymous namespace)::runMixxx (args=<optimized out>, pApp=0x7fffffffdbc0) at ../src/main.cpp:41
#8  main(int, char**) (argc=<optimized out>, argv=0x7fffffffddd8) at ../src/main.cpp:189
(gdb) bt full
#0  DlgPreferences::~DlgPreferences() (this=0x9c1d060, __in_chrg=<optimized out>) at ../src/preferences/dialog/dlgpreferences.cpp:297
#1  0x000000000059394a in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release() (this=0x9c1d050)
    at /usr/include/c++/11/bits/shared_ptr_base.h:168
#2  std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release() (this=0x9c1d050) at /usr/include/c++/11/bits/shared_ptr_base.h:161
#3  std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count() (this=0x7fffffffdc48, __in_chrg=<optimized out>)
    at /usr/include/c++/11/bits/shared_ptr_base.h:705
#4  std::__shared_ptr<DlgPreferences, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr() (this=0x7fffffffdc40, __in_chrg=<optimized out>)
    at /usr/include/c++/11/bits/shared_ptr_base.h:1154
#5  std::shared_ptr<DlgPreferences>::~shared_ptr() (this=0x7fffffffdc40, __in_chrg=<optimized out>) at /usr/include/c++/11/bits/shared_ptr.h:122
#6  mixxx::qml::QmlApplication::~QmlApplication() (this=0x7fffffffdc00, __in_chrg=<optimized out>) at ../src/qml/qmlapplication.h:19
#7  0x00000000005740c9 in (anonymous namespace)::runMixxx (args=<optimized out>, pApp=0x7fffffffdbc0) at ../src/main.cpp:41
```

Be-ing commented on src/qml/qmlapplication.cpp:
```diff
@@ -58,6 +68,25 @@ QmlApplication::QmlApplication(
+            1,
+            "PreferencesDialog",
+            lambda_to_singleton_type_factory_ptr(
+                    [this](QQmlEngine* pEngine,
```
I also tried making the DlgPreferences instance a local variable of the QmlApplication constructor and a local variable within this lambda to test if capturing `this` was causing the problem, but neither of those solved the problem.

Be-ing commented:
Well this is interesting. I replaced the DlgPreferences instance with a simple QDialogButtonBox and reproduced the issue of QmlApplication shutting down when the widget is destroyed.

Be-ing commented:
I figured out the problem. QApplication's event loop [shuts down the QApplication when the last QWidget window is closed](https://code.woboq.org/qt5/qtbase/src/widgets/kernel/qapplication.cpp.html#2861). I confirmed this by showing a dummy QDialogButtonBox then opening the preferences from QML. In that case, I could close the preferences without shutting down the application but the application shut down after closing both the preferences and the dummy QDialogButtonBox.

I attempted changing MixxxApplication to subclass QGuiApplication instead of QApplication, but then it is not possible to use any QWidgets at all; at runtime console output shows:
```
QWidget: Cannot create a QWidget without QApplication
```

So I think we need to filter out the erroneous QEvent::Close event in MixxxApplication when the QQmlApplicationEngine is still showing to avoid needing to rewrite the whole preferences in QML immediately.

Holzhaus commented:
Just setting this to false doesn't work?

Be-ing commented:
Ha! That looks like a simpler solution, thanks.

Be-ing commented:
Adding
```
    app->setQuitOnLastWindowClosed(false);
    app->setQuitLockEnabled(false);
```
to the QmlApplication constructor does not solve the problem.

Be-ing commented:
I found a one line fix:
```
m_pDlgPreferences->setAttribute(Qt::WA_QuitOnClose, false);
```
Now to deal with the hang in ~ControllerManager....

Be-ing commented:
Well this is interesting. I found a workaround for the hack to include the uic generated header for the mixxx executable CMake target. Instead of making DlgPreferences a std::shared_ptr member of QmlApplication, I changed it to a local std::shared_ptr in the QmlApplication constructor. That allowed to move the `#include "preferences/dialog/dlgpreferences.h" to the cpp file which is not `#include`d in the mixxx executable target. Then, magically, ~ControllerManager stopped hanging on shutdown. Go figure.

Now there are more errors on shutdown to address.

Be-ing commented:
Oof, it seems that DlgPreferences is outliving CoreServices...

Be-ing commented:
> Then, magically, ~ControllerManager stopped hanging on shutdown.

That is because ~ControllerManager is not getting called because ~CoreServices is running before ~DlgPreferences...

Be-ing commented:
I think I have figured out how to get the proper lifetime for DlgPreferences and added extensive comments in the QmlApplication constructor explaining this. Now, why is ~ControllerManager hanging...

Be-ing commented:
The issue with ~ControllerManager hanging seems to be because CoreServices gets destroyed after the Qt event loop stops processing events. That seems to happen because std::shared_ptr references are captured by the lambdas for qmlRegisterSingletonType which outlive QmlApplication.

Be-ing commented:
Confirmed. The qmlRegisterSingletonType lambdas must not retain ownership of CoreServices or its manager members. Maybe std::weak_ptr can help.

Be-ing commented:
Yup, the solution was to capture std::weak_ptrs in the lambdas and convert them to std::shared_ptrs when the lambdas are executed by the QQmlEngine to instantiate the QML proxy objects.

Be-ing commented:
Ready for review.

Be-ing committed:
ErrorDialogHandler: set WA_QuitOnClose to false

to prevent closing QWidget error dialogs from shutting down Mixxx
when using the QQmlApplication GUI

Be-ing committed:
QmlApplication: capture weak pointers in lambdas

Be-ing committed:
add QML button to show DlgPreferences

Swiftb0y commented on src/errordialoghandler.cpp:
```diff
@@ -162,6 +162,7 @@ void ErrorDialogHandler::errorDialog(ErrorDialogProperties* pProps) {
 
     // This deletes the msgBox automatically, avoiding a memory leak
     pMsgBox->setAttribute(Qt::WA_DeleteOnClose, true);
+    pMsgBox->setAttribute(Qt::WA_QuitOnClose, false);
```
I'd document why this required.

Be-ing replied:
comment added

Swiftb0y commented on src/preferences/dialog/dlgprefinterface.cpp:
```diff
@@ -67,11 +67,6 @@ DlgPrefInterface::DlgPrefInterface(
     double unscaledDevicePixelRatio = m_dDevicePixelRatio / initialScaleFactor;
     m_minScaleFactor = 1 / unscaledDevicePixelRatio;
 
-    VERIFY_OR_DEBUG_ASSERT(m_pSkin != nullptr) {
```
Have you made sure we don't assume that `m_pSkin != nullptr` anywhere else?

Be-ing replied:
yes

Swiftb0y commented on res/qml/main.qml:
```diff
@@ -64,6 +65,17 @@ Rectangle {
+
+                    text: "Prefs"
+                    activeColor: Theme.white
+                    checkable: true
+                    onClicked: {
+                        Mixxx.PreferencesDialog.show();
+                    }
```
Not sure if it makes sense for the button to be checkable, when nothing happens when unchecking.

Be-ing replied:
I tried to add a Q_PROPERTY to the QmlDlgPreferencesProxy class indicating the state of the dialog but unfortunately there is no notifyChanged callback for QWidget's `visible` property.

Be-ing replied:
Simply removing `checkable: true` works well.

Swiftb0y reviewed (commented):
LGTM, but I'd like to have a second opinion before merging if possible.

uklotzde commented on src/preferences/dialog/dlgpreferences.cpp:
```diff
@@ -152,17 +152,20 @@ DlgPreferences::DlgPreferences(
-            this,
-            &DlgPreferences::reloadUserInterface,
-            Qt::DirectConnection);
+    // ugly proxy for determining whether this is being instatiated for QML or legacy QWidgets GUI
```
typo

Be-ing replied:
fixed

Be-ing commented:
Merge?

Holzhaus commented:
This PR leads to 2 major regressions:

1. Closing any popup closes Mixxx. E.g. if a track fails to load, a popup will open. If you close it, Mixxx shuts down.
2. It's not possible to use the developer tools anymore, which makes QML skin development extremely tedious.

Be-ing commented:
>  Closing any popup closes Mixxx. E.g. if a track fails to load, a popup will open. If you close it, Mixxx shuts down.

Qt::WA_QuitOnClose needs to be set to false for every QWidget window: https://github.com/mixxxdj/mixxx/pull/4327/files#diff-2768d0d13eda8f9b18b5ce2c75faed57cb48c7f8463612e674455561533234f7R63
